File: 2022/4_Camp_cleanup/part_2.rb
input =
  File
    .open('input.txt')
    .read
    .split("\n")
    .map do |line|
      string_ranges = line.split(',')
      string_ranges.map do |range|
        range_numbers = range.split('-').map { |string| string.to_i }
        { 'min' => range_numbers[0], 'max' => range_numbers[1] }
      end
    end

def ranges_overlap(pair)
  (pair[0]['min'] >= pair[1]['min'] && pair[0]['min'] <= pair[1]['max']) ||
    (pair[0]['max'] >= pair[1]['min'] && pair[0]['max'] <= pair[1]['max']) ||
    (pair[1]['min'] >= pair[0]['min'] && pair[1]['min'] <= pair[0]['max']) ||
    (pair[1]['max'] >= pair[0]['min'] && pair[1]['max'] <= pair[0]['max'])
end

counter = 0

for pair in input
  counter += 1 if ranges_overlap(pair)
end

puts counter

File: 2022/4_Camp_cleanup/part_2.ts
import { readFileSync } from 'fs'

const input: Range[][] = readFileSync('input.txt', 'utf-8')
  .split('\n')
  .map((line) =>
    line.split(',').map((pair) => {
      const stringPairs = pair.split('-')
      return {
        min: parseInt(stringPairs[0]),
        max: parseInt(stringPairs[1]),
      }
    })
  )

type Range = {
  min: number
  max: number
}

let counter = 0
// iterate over each pairings
// check if they overlap
// if they are then increment counter

input.forEach((pairing) => {
  if (pairingsOverlap(pairing)) {
    counter++
  }
})

function pairingsOverlap(pairing: Range[]): boolean {
  return (
    (pairing[0].min >= pairing[1].min && pairing[0].min <= pairing[1].max) ||
    (pairing[0].max >= pairing[1].min && pairing[0].max <= pairing[1].max) ||
    (pairing[1].min >= pairing[0].min && pairing[1].min <= pairing[0].max) ||
    (pairing[1].max >= pairing[0].min && pairing[1].max <= pairing[0].max)
  )
}

console.log(counter)
